### PR TITLE
Auto ref counting, refactoring of some OOP functions, new SQF-VM, fixes.

### PR DESCRIPTION
**Pull request by billw2012**

### CODE REVIEW

File: Project_0.Altis/OOP_Light/OOP_Light.h [hunk 1]
```diff
@@ -290,31 +307,106 @@ private _classNameStr = OBJECT_PARENT_CLASS_STR(_objNameStr);
 #define REMOTE_EXEC_CALL_STATIC_METHOD(classNameStr, methodNameStr, extraParams, targets, JIP) ([classNameStr] + extraParams) remoteExecCall [CLASS_METHOD_NAME_STR(classNameStr, methodNameStr), targets, JIP];
 #endif
 
+// ----------------------------------------
+// |         A T T R I B U T E S          |
+// ----------------------------------------
+
+#define ATTR_REFCOUNTED 1
+#define ATTR_SERIALIZABLE 2
+#define ATTR_USERBASE 1000
```
Comment (Sparker95) [inline Project_0.Altis/OOP_Light/OOP_Light.h]: Should be enough

File: Project_0.Altis/OOP_Light/OOP_Light.h [hunk 2]
```diff
@@ -350,10 +442,7 @@ private _oop_parents = []; \
 private _oop_methodList = []; \
 private _oop_newMethodList = []; \
 if (baseClassNameStr != "") then { \
-	if (!([baseClassNameStr, __FILE__, __LINE__] call OOP_assert_class)) then { \
-		missionNamespace setVariable [OOP_ERROR_DEBRIEFING_SECTION_VAR_NAME_STR, format ["Class %1 is not defined. File: %2", baseClassNameStr, __FILE__]]; \
-		endMission OOP_ERROR_DEBRIEFING_CLASS_NAME_STR; \
-	}; \
+	if (!([baseClassNameStr, __FILE__, __LINE__] call OOP_assert_class)) then {throw "invalid base class"}; \
```
Comment (Sparker95) [inline Project_0.Altis/OOP_Light/OOP_Light.h]: Who is catching it?

**Comment by Sparker95** [comment]
Overall looks very good. I only don't understand how the CLASS macro works with the throw operator. It's not inside a try block, so what's the point?

File: Project_0.Altis/OOP_Light/OOP_Light.h (hunk 2 above)
Comment (billw2012) [reply Project_0.Altis/OOP_Light/OOP_Light.h]: Nobody, but it does exit the script immediately. Assert does the error surfacing itself right? We could certainly put catch block in init that would catch all of these if we wanted though.

**Pull request merged**